plugins/extraction/table.ml

```diff
@@ -842,6 +930,35 @@ let extract_constant_inline inline r ids s =
         Lib.add_leaf (in_customs (g,ids,s))
     | _ -> error_constant ?loc:r.CAst.loc g
 
+(* const_name : qualid -> params : string list -> replacement : string*)
+let extract_constant_foreign r ids s =
+  if lang () != Ocaml then
+      CErrors.user_err (Pp.str "Extract Foreign Constant is supported only for OCaml extraction");
+  check_inside_section ();
+  (* Get the global id of the constant *)
+  let g = Smartlocate.global_with_alias r in
+  match g with
+    (* It is indeed a constant reference. *)
+    | GlobRef.ConstRef kn ->
+        let env = Global.env () in
+        let typ, _ = Typeops.type_of_global_in_context env (GlobRef.ConstRef kn) in
+        let typ = Reduction.whd_all env typ in
+        if Reduction.is_arity env typ
+          then begin
+            let nargs = Hook.get use_type_scheme_nb_args env typ in
+            if not (Int.equal (List.length ids) nargs) then error_axiom_scheme ?loc:r.CAst.loc g nargs
+          end;
+        (* We have the global id and the args now. *)
+        (* Add g to foreign extractions.*)
+        Lib.add_leaf (foreign_extraction [g]);
+        (* Also, if g already is a custom, make sure the user is informed by throwing the error *)
+        if (is_custom g)
+        then CErrors.user_err ((str "The term ") ++ safe_pr_long_global g ++ (str " is already defined as custom constant."))
+        else ();
```

Comment:
The `else ()` branch is useless.

Also, I do not quite understand the logic here. I don't think there is anything wrong with redefining as foreign a function that is already custom. What might be wrong, however, is to redefine as foreign a function that is already marked for inlining, and also to redefine as custom a function that is already marked as foreign (which should thus be tested in the other function).

Reply:
Okay, that makes sense. I adopted the code accordingly. Will go upstream soon